includes/utils/class-amp-mutation-utils.php

```diff
@@ -0,0 +1,325 @@
+<?php
+/**
+ * Class AMP_Mutation_Utils
+ *
+ * @package AMP
+ */
+
+/**
+ * Class AMP_Mutation_Utils
+ *
+ * @since 0.7
+ */
+class AMP_Mutation_Utils {
+
+	/**
+	 * The argument if an attribute was removed.
+	 *
+	 * @const array.
+	 */
+	const ATTRIBUTE_REMOVED = 'removed_attr';
+
+	/**
+	 * The argument if a node was removed.
+	 *
+	 * @const array.
+	 */
+	const NODE_REMOVED = 'removed';
+
+	/**
+	 * Key for the markup value in the REST API endpoint.
+	 *
+	 * @const string.
+	 */
+	const MARKUP_KEY = 'markup';
+
+	/**
+	 * Key for the error value in the response.
+	 *
+	 * @const string.
+	 */
+	const ERROR_KEY = 'has_error';
+
+	/**
+	 * Key of the AMP error query var.
+	 *
+	 * @const string.
+	 */
+	const ERROR_QUERY_KEY = 'amp_error';
+
+	/**
+	 * Query arg value if there is an AMP error in the post content.
+	 *
+	 * @const string.
+	 */
+	const ERROR_QUERY_VALUE = '1';
+
+	/**
+	 * The attributes that the sanitizer removed.
+	 *
+	 * @var array.
+	 */
+	public static $removed_attributes;
+
+	/**
+	 * The nodes that the sanitizer removed.
+	 *
+	 * @var array.
+	 */
+	public static $removed_nodes;
+
+	/**
+	 * Add the actions.
+	 *
+	 * @return void.
+	 */
+	public static function init() {
+		add_action( 'rest_api_init', array( __CLASS__, 'amp_rest_validation' ) );
+		add_action( 'save_post', array( __CLASS__, 'validate_content' ), 10, 2 );
+		add_action( 'edit_form_top', array( __CLASS__, 'display_error' ) );
+	}
+
+	/**
+	 * Tracks when a sanitizer removes an attribute or node.
+	 *
+	 * @param DOMNode|DOMElement $node The node in which there was a removal.
+	 * @param string             $removal_type The removal: 'removed_attr' for an attribute, or 'removed' for a node or element.
+	 * @param string             $attr_name The name of the attribute removed (optional).
+	 * @return void.
+	 */
+	public static function track_removed( $node, $removal_type, $attr_name = null ) {
+		if ( ( self::ATTRIBUTE_REMOVED === $removal_type ) && isset( $attr_name ) ) { // phpcs:ignore WordPress.NamingConventions.ValidVariableName.NotSnakeCaseMemberVar
+			self::$removed_attributes = self::increment_count( self::$removed_attributes, $attr_name );
+		} elseif ( ( self::NODE_REMOVED === $removal_type ) && isset( $node->nodeName ) ) {
+			self::$removed_nodes = self::increment_count( self::$removed_nodes, $node->nodeName ); // phpcs:ignore WordPress.NamingConventions.ValidVariableName.NotSnakeCaseMemberVar
+		}
+	}
+
+	/**
+	 * Tracks when a sanitizer removes an attribute or node.
+	 *
+	 * @param array  $histogram The count of attributes or nodes removed.
+	 * @param string $key The attribute or node name removed.
```

Comment:
It is a wonder Travis didn't pickup alignment comments miss alignment, it might only sniff the variables alignment actually.

Reply:
Should this also align the comments, like:
```
@param array  $histogram The count of attributes or nodes removed.
@param string $key       The attribute or node name removed.
```

Reply:
Yes

Reply:
This [commit](https://github.com/Automattic/amp-wp/commit/cd910ff648356a9d454ca4c3148d6f9935544260) aligns the spacing of the comments. The pre-commit hook blocked aligning this in [two](https://github.com/Automattic/amp-wp/pull/912/files#diff-7d234989187d87a372e2aa6aa1299a71R245) [places](https://github.com/Automattic/amp-wp/pull/912/files#diff-7d234989187d87a372e2aa6aa1299a71R313), possibly because there was only one `@param` and a `@return`.